protocol/kafka_confluent/v2/protocol.go

```
@@ -0,0 +1,225 @@
+	if err != nil {
+		return err
+	}
+
+	err = p.producer.Produce(kafkaMsg, p.producerDeliveryChan)
+	if err != nil {
+		return err
+	}
+	e := <-p.producerDeliveryChan
+	m := e.(*kafka.Message)
+	if m.TopicPartition.Error != nil {
+		return m.TopicPartition.Error
+	}
+	return nil
+}
+
+func (p *Protocol) OpenInbound(ctx context.Context) error {
+	if p.consumer == nil {
+		return fmt.Errorf("the consumer client must be set")
+	}
+	if p.consumerTopics == nil {
+		return fmt.Errorf("the consumer topics must be set")
+	}
+
+	p.consumerMux.Lock()
+	defer p.consumerMux.Unlock()
+	logger := cecontext.LoggerFrom(ctx)
+
+	// Query committed offsets for each partition
+	if positions := TopicPartitionOffsetsFrom(ctx); positions != nil {
+		if err := p.consumer.Assign(positions); err != nil {
+			return err
+		}
+	}
+
+	logger.Infof("Subscribing to topics: %v", p.consumerTopics)
+	err := p.consumer.SubscribeTopics(p.consumerTopics, p.consumerRebalanceCb)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if !p.consumer.IsClosed() {
+			logger.Infof("Closing consumer %v", p.consumerTopics)
+			if err = p.consumer.Close(); err != nil {
+				logger.Errorf("failed to close the consumer: %v", err)
+			}
+		}
+	}()
+
+	for !p.consumer.IsClosed() {
+		select {
+		case <-ctx.Done():
+			logger.Debugf("The consumer context is canceled!")
+			return nil
+		default:
+			ev := p.consumer.Poll(p.consumerPollTimeout)
+			if ev == nil {
+				continue
+			}
+			switch e := ev.(type) {
+			case *kafka.Message:
+				p.incoming <- e
+			case kafka.Error:
+				// Errors should generally be considered informational, the client will try to automatically recover.
+				// But in here, we choose to terminate the application if all brokers are down.
+				logger.Infof("Error %v: %v", e.Code(), e)
+				if p.consumerErrorHandler != nil {
+					p.consumerErrorHandler(e)
+				}
+				if e.Code() == kafka.ErrAllBrokersDown {
+					logger.Debugf("All broker connections are down!")
```

so let's make this `Errorf` then to not hide this error when debugging is off.

---

Done.